platform-core-module/src/test/java/com/softicar/platform/core/module/program/execution/cleanup/ProgramExecutionDeleterTest.java

```diff
@@ -0,0 +1,113 @@
+	private final Day referenceDay = CURRENT_DAYTIME.getDay();
+	private final SetMap<AGProgram, AGProgramExecution> expectedRetainedProgramExecutionsMap = new SetMap<>();
+
+	public ProgramExecutionDeleterTest() {
+
+		// Suppress lower-level log output under test
+		LogLevel.ERROR.set();
+
+		// Use of special clock for JUnit tests because ProgramExecutionDeleter uses the current day to load the to be deleted program executions
+		testClock.setInstant(Instant.ofEpochMilli(CURRENT_DAYTIME.toMillis()));
+
+		for (AGProgram program: insertPrograms()) {
+			insertExecutionsOfProgram(program);
+		}
+	}
+
+	private void insertExecutionsOfProgram(AGProgram program) {
+
+		for (Day terminatedAtDay: new DateItemRange<>(referenceDay.getRelative(-4), referenceDay)) {
+			AGProgramExecution execution = insertProgramExecutionWithLog(program, terminatedAtDay);
+			putExecutionOnExpectedRetainedProgramExecutionsMapIfNecessary(program, terminatedAtDay, execution);
+		}
+	}
+
+	private void putExecutionOnExpectedRetainedProgramExecutionsMapIfNecessary(AGProgram program, Day terminatedAtDay, AGProgramExecution execution) {
+
+		Day minimalRetainedDay = referenceDay.getRelative(-program.getRetentionDaysOfExecutions());
+		if (terminatedAtDay.compareTo(minimalRetainedDay) >= 0) {
+			expectedRetainedProgramExecutionsMap.addToSet(program, execution);
+		}
+	}
+
+	private List<AGProgram> insertPrograms() {
+
+		List<AGProgram> programs = new ArrayList<>();
+		for (int retentionDaysOfExecutions = 0; retentionDaysOfExecutions < 3; retentionDaysOfExecutions++) {
+			programs.add(insertProgram(retentionDaysOfExecutions));
+		}
+		return programs;
+	}
+
+	private AGProgram insertProgram(int retentionDaysOfExecutions) {
+
+		AGUuid uuid = insertRandomUuid();
+		return new AGProgram()//
+			.setProgramUuid(uuid)
+			.setRetentionDaysOfExecutions(retentionDaysOfExecutions)
+			.save();
+	}
+
+	private AGUuid insertRandomUuid() {
+
+		UUID uuid = UUID.randomUUID();
+		byte[] uuidBytes = UuidBytes.asBytes(uuid);
+		return new AGUuid().setUuidString(uuid.toString()).setUuidBytes(uuidBytes).save();
+	}
+
+	private AGProgramExecution insertProgramExecutionWithLog(AGProgram program, Day terminatedAtDay) {
+
+		AGProgramExecution programExecution = new AGProgramExecution()//
+			.setProgramUuid(program.getProgramUuid())
+			.setTerminatedAt(DayTime.fromDayAndSeconds(terminatedAtDay, 0))
+			.setQueuedBy(CurrentUser.get())
+			.save();
+		insertLogRecord(programExecution);
+		return programExecution;
+	}
+
+	private void insertLogRecord(AGProgramExecution programExecution) {
+
+		AGTransaction transaction = new AGTransaction().setAt(DayTime.now()).setByToCurrentUser().save();
+		AGProgramExecutionLog.TABLE.getOrCreate(new Tuple2<>(programExecution, transaction));
+	}
+
+	@Test
+	public void testDelete() {
+
+		new ProgramExecutionDeleter(0).delete();
+		assertEquals(expectedRetainedProgramExecutionsMap, loadProgramExecutions());
```

Comment:
Assertions should be more explicit.
To me, the content of `expectedRetainedProgramExecutionsMap` at this point is totally unclear... I need to read and understand the whole test setup to comprehend the expectation in this assertion.
I think this needs to be simplified, and/or made more explicit - e.g. by explicitly defining the expected executions in this test method.

Reply:
I improved the test: It should be easier to understand now. I added more test methods to validate different scenarios.
Please have a look.